eta/core/datasets.py

```diff
@@ -1071,98 +1069,112 @@ class LazyLabeledVideoEntry(LazyLabeledDataEntry):
+                                                    total_frame_count, duration)
+        self.end = etav.frame_number_to_timestamp(end_frame, total_frame_count,
+                                                  duration)
         self.start_frame = start_frame
         self.end_frame = end_frame
         self.total_frames = total_frame_count
 
-
     def get_labels(self):
         if self.label_obj:
             return self.label_obj
-        #@TODO - Handle file reading from ANY SOURCE!!! (path may not be local)
+        # @TODO - Handle file reading from ANY SOURCE!!! (path may not be local)
         return etav.VideoLabels.from_json(self.label_path)
 
 
-
-class LabeledDatasetBuilder(object):
+class DatasetTransformer(object):
     '''
-    This object builds a LabeledDataset with optional transformations such as
-    sampling, filtering by schema, and balance.
+    Object responsible for transforming TransformableDatasets
+    Basically, strategy pattern
     '''
 
-    def __init__(self, index=LazyTransformableDataset()):
-        self.index = index
-        self.schema = None
-        self.balance_attr = None
-        self.sample_rate = None
+    def transform(self, input):
+        ''' Transform a TransformableDataset
+        Args:
+            input (LazyTransformableDataset): the input dataset
 
+        Returns:
+        '''
+        return input
 
-    def sample(self, k):
-        self.sample_rate = k
-        return self
 
-    def filter(self, schema):
-        self.schema = schema
-        return self
+class Sampler(DatasetTransformer):
+    '''
+    Randomly sample the number of entries in the dataset to some number k
+    '''
 
-    def balance(self, attribute):
-        self.balance_attr = attribute
-        return self
+    def __init__(self, k):
+        super(Sampler, self).__init__()
+        self.k = k
 
-    def _balance(self):
-        #@TODO
-        if self.balance_attr:
-            return
+    def transform(self, input):
+        input.set_entries(random.sample(input.get_entries(), self.k))
+        return input
 
-    def _filter(self):
-        #@TODO
-        if self.schema:
-            return
 
-    def _sample(self):
-        self.index.set_entries(
-            random.sample(self.index.get_entries(), self.sample_rate))
+class Balancer(DatasetTransformer):
+    '''
+    Balance the number of entries in the dataset by values in some categorical
+    Attribute, as provided on construction.
+    '''
 
-    def build(self):
-        #@TODO these might return stuff?
-        self._sample()
-        self._filter()
-        self._balance()
+    def __init__(self, attribute):
+        super(Balancer, self).__init__()
+        self.attr = attribute
 
-        #@TODO build me!
-        return LabeledDataset(None)
+    def transform(self, input):
+        # @TODO implement Balancing!!
+        return input
 
 
-class LabeledVideoDatasetBuilder(LabeledDatasetBuilder):
+class Filter(DatasetTransformer):
```

Comment:
small comment on naming: `Filter` is so generic. `DatasetFilter` would be preferable. Same for other classes

Reply:
Yes - will fix that